### PR TITLE
[MDCNavigationBar] Exposing typography API.

### PR DESCRIPTION
**Pull request by mohammadcazig**
Title Font should be customizable, however font size should be fixed for navigation bar.
Pivotal: #155533133

### CODE REVIEW

File: components/NavigationBar/src/MDCNavigationBar.h [hunk 1]
```diff
@@ -98,6 +98,13 @@ IB_DESIGNABLE
  */
 @property(nonatomic, strong, nullable) UIView *titleView;
 
+/**
+ The font applied to the title of navigation bar. If titleTextAttributes is not nil, it will
```
Comment (jverkoey) [inline components/NavigationBar/src/MDCNavigationBar.h]: What happens if the attributes are provided but don’t have a font specified?

File: components/NavigationBar/src/MDCNavigationBar.m [hunk 2]
```diff
@@ -261,6 +263,14 @@ - (void)encodeWithCoder:(NSCoder *)aCoder {
   [aCoder encodeInteger:self.titleAlignment forKey:MDCNavigationBarTitleAlignmentKey];
 }
 
+- (void)setTitleFont:(UIFont *)titleFont {
+  _titleFont = [titleFont fontWithSize:kTitleFontSize];
+  if (!_titleFont) {
+    _titleFont = [MDCTypography titleFont];
```
Comment (jverkoey) [inline components/NavigationBar/src/MDCNavigationBar.m:269]: We should add a follow up task here to capture the fact that mdc typography can be used to change the font size to a different size than 20. I see that you made the change here to be compatible with existing usage, but if our intent is to enforce a size of 20 this line will allow clients to deviate from that.

File: components/NavigationBar/src/MDCNavigationBar.h (hunk 1 above)
Comment (mohammadcazig) [reply components/NavigationBar/src/MDCNavigationBar.h]: We fall into default behavior of UILabel.

File: components/NavigationBar/src/MDCNavigationBar.m (hunk 2 above)
Comment (mohammadcazig) [reply components/NavigationBar/src/MDCNavigationBar.m:269]: I will add an issue, but just to be clear this was already possible by setting text attributes.

File: components/NavigationBar/src/MDCNavigationBar.h (hunk 1 above)
Comment (jverkoey) [reply components/NavigationBar/src/MDCNavigationBar.h]: It may be beneficial to document that this API is ultimately hooking in to a UILabel's font property so that that behavior is well-understood by the client.

With that addition, the "If titleTextAttributes is not nil, it will..." line can probably be removed.

**Review by jverkoey — APPROVED** [review]

File: components/NavigationBar/src/MDCNavigationBar.h (hunk 1 above)
Comment (jverkoey) [reply components/NavigationBar/src/MDCNavigationBar.h]: Also document what happens when this property is set to nil (because it is null_resettable).

**Review by jverkoey — DISMISSED** [review]
Docs requests.

File: components/NavigationBar/src/MDCNavigationBar.h [hunk 3]
```diff
@@ -98,6 +98,12 @@ IB_DESIGNABLE
  */
 @property(nonatomic, strong, nullable) UIView *titleView;
 
+/**
+ The font applied to the title of navigation bar. This font is set through UILabel Font property.
+ Font size is enforced to 20.
```
Comment (jverkoey) [inline components/NavigationBar/src/MDCNavigationBar.h]: My previous comment got lost unfortunately...

We need to document what happens when this property is set to nil because the property is `null_resettable`.

**Review by jverkoey — DISMISSED** [review]

File: components/NavigationBar/src/MDCNavigationBar.h (hunk 3 above)
Comment (romoore) [inline components/NavigationBar/src/MDCNavigationBar.h]: > Font size is enforced to 20.

Just to clarify, does this mean that no matter what UIFont I provide, it will resize the font to 20 points?

File: components/NavigationBar/src/MDCNavigationBar.h [hunk 4]
```diff
@@ -98,6 +98,12 @@ IB_DESIGNABLE
  */
 @property(nonatomic, strong, nullable) UIView *titleView;
 
+/**
+ The font applied to the title of navigation bar. This font is set through UILabel Font property.
```
Comment (romoore) [inline components/NavigationBar/src/MDCNavigationBar.h]: Why do we have a separate API if clients should "set through UILabel Font property"? Can we not just expect them to set that property instead?

File: components/NavigationBar/src/MDCNavigationBar.h (hunk 3 above)
Comment (mohammadcazig) [reply components/NavigationBar/src/MDCNavigationBar.h]: Correct.

Comment (jverkoey) [reply components/NavigationBar/src/MDCNavigationBar.h]: I've filed https://www.pivotaltracker.com/story/show/156482550 to discuss this further. I also agree that the intention to fix to a specific font size is a bit surprising.

File: components/NavigationBar/src/MDCNavigationBar.h [hunk 5]
```diff
@@ -99,18 +99,11 @@ IB_DESIGNABLE
 @property(nonatomic, strong, nullable) UIView *titleView;
 
 /**
- Display attributes for the titleView's title text.
-
- Setting this property will render an NSAttributedString with the assigned attributes across the
- entire text.
+ The font applied to the title of navigation bar.
+ Font size is enforced to 20.
```
Comment (romoore) [inline components/NavigationBar/src/MDCNavigationBar.h]: Are we sure the font size enforcement is absolutely necessary? Can we just recommend that clients stick to a 20-point (or similarly-sized) font if they use a custom font?

File: components/NavigationBar/src/MDCNavigationBar.m [hunk 6]
```diff
@@ -261,6 +263,14 @@ - (void)encodeWithCoder:(NSCoder *)aCoder {
   [aCoder encodeInteger:self.titleAlignment forKey:MDCNavigationBarTitleAlignmentKey];
 }
 
+- (void)setTitleFont:(UIFont *)titleFont {
```
Comment (romoore) [inline components/NavigationBar/src/MDCNavigationBar.m:266]: Please add a unit test for this method.

**Review by romoore — DISMISSED** [review]

File: components/NavigationBar/src/MDCNavigationBar.h (hunk 5 above)
Comment (mohammadcazig) [reply components/NavigationBar/src/MDCNavigationBar.h]: I actually don't have an strong opinion on this. @ianegordon do you think we should allow users to set any font, leaving it to them to make sure it looks correct with their font?

File: components/NavigationBar/src/MDCNavigationBar.m (hunk 6 above)
Comment (mohammadcazig) [reply components/NavigationBar/src/MDCNavigationBar.m:266]: Done.

File: components/NavigationBar/src/MDCNavigationBar.h (hunk 5 above)
Comment (jverkoey) [reply components/NavigationBar/src/MDCNavigationBar.h]: Comments keep getting lost, so reposting here:

I've filed https://www.pivotaltracker.com/story/show/156482550 to discuss this further. I also agree that the intention to fix to a specific font size is a bit surprising, but this PR is not changing the behavior. Any fix or change to that behavior can and should be done in a follow-up PR.

File: components/NavigationBar/src/MDCNavigationBar.h [hunk 7]
```diff
@@ -99,18 +99,11 @@ IB_DESIGNABLE
...
+ Default is MDCTypography's titleFont.
```
Comment (jverkoey) [inline components/NavigationBar/src/MDCNavigationBar.h]: Please document what happens when this property is set to nil.

File: components/NavigationBar/src/MDCNavigationBar.h [hunk 8]
```diff
@@ -193,7 +186,25 @@ IB_DESIGNABLE
...
+ */
+#if defined(__IPHONE_11_0) && (__IPHONE_OS_VERSION_MAX_ALLOWED >= __IPHONE_11_0)
+@property(nonatomic, copy, nullable)
+    NSDictionary<NSAttributedStringKey, id> *titleTextAttributes UI_APPEARANCE_SELECTOR
+    __deprecated_msg("Use titleFont instead");
+#else
+@property(nonatomic, copy, nullable)
+    NSDictionary<NSString *, id> *titleTextAttributes UI_APPEARANCE_SELECTOR
+    __deprecated_msg("Use titleFont instead");
+#endif
```
Comment (jverkoey) [inline components/NavigationBar/src/MDCNavigationBar.h]: This API is presently being used by internal clients - please follow the [deprecation policy](https://github.com/material-components/material-components-ios/blob/develop/contributing/deprecation_policy.md) here.

tl;dr: Only deprecate APIs once you've defined a deprecation plan and timeline. For now, please revert these deprecations in this PR.

File: components/NavigationBar/src/MDCNavigationBar.h [hunk 9]
```diff
@@ -99,18 +99,11 @@ IB_DESIGNABLE
...
+ Default is MDCTypography's titleFont.
  */
-#if defined(__IPHONE_11_0) && (__IPHONE_OS_VERSION_MAX_ALLOWED >= __IPHONE_11_0)
-@property(nonatomic, copy, nullable)
-    NSDictionary<NSAttributedStringKey, id> *titleTextAttributes UI_APPEARANCE_SELECTOR;
-#else
-@property(nonatomic, copy, nullable)
-    NSDictionary<NSString *, id> *titleTextAttributes UI_APPEARANCE_SELECTOR;
-#endif
+@property (nonatomic, strong, null_resettable) UIFont *titleFont;
```
Comment (jverkoey) [inline components/NavigationBar/src/MDCNavigationBar.h]: Also document how this property relates to `titleTextAttributes`. Notably, call out:

- `titleTextAttributes` will be deprecated in a future release and you should prefer `titleFont` instead.
- If a client sets both properties, document what the expected behavior is.

**Review by jverkoey — DISMISSED** [review]
More documentation requests.

File: components/NavigationBar/src/MDCNavigationBar.h (hunk 7 above)
Comment (mohammadcazig) [reply components/NavigationBar/src/MDCNavigationBar.h]: it's null_resetable to default, MDCTypography titleFont is nonnull.

File: components/NavigationBar/src/MDCNavigationBar.h (hunk 8 above)
Comment (mohammadcazig) [reply components/NavigationBar/src/MDCNavigationBar.h]: Alright I created an issue. #3224

File: components/NavigationBar/src/MDCNavigationBar.h (hunk 7 above)
Comment (jverkoey) [reply components/NavigationBar/src/MDCNavigationBar.h]: There are two things we need to document here:

1. What the default, initial value is.
2. What happens when we set the property to nil.

File: components/NavigationBar/src/MDCNavigationBar.h [hunk 10]
```diff
@@ -98,9 +98,17 @@ IB_DESIGNABLE
  */
 @property(nonatomic, strong, nullable) UIView *titleView;
 
+/**
+ The font applied to the title of navigation bar.
+ Font size is enforced to 20.
+ Default is MDCTypography's titleFont.
+ */
+@property (nonatomic, strong, null_resettable) UIFont *titleFont;
```
Comment (jverkoey) [inline components/NavigationBar/src/MDCNavigationBar.h]: Also document how this property relates to `titleTextAttributes`. Notably, call out:

- `titleTextAttributes` will be deprecated in a future release and you should prefer `titleFont` instead.
- If a client sets both properties, document what the expected behavior is.

**Review by jverkoey — DISMISSED** [review]
Please address the outstanding comments regarding documentation.

File: components/NavigationBar/src/MDCNavigationBar.h (hunk 10 above)
Comment (mohammadcazig) [reply components/NavigationBar/src/MDCNavigationBar.h]: I am sorry, everything kinda got mixed up. 
Done.

File: components/NavigationBar/src/MDCNavigationBar.h [hunk 11]
```diff
@@ -98,11 +98,21 @@ IB_DESIGNABLE
  */
 @property(nonatomic, strong, nullable) UIView *titleView;
 
+/**
+ The font applied to the title of navigation bar.
+ Font size is enforced to 20.
+ Default is MDCTypography's titleFont.
+ Note that Font attribute will of titleTextAttributes take precedence over titleFont property.
```
Comment (jverkoey) [inline components/NavigationBar/src/MDCNavigationBar.h]: minor grammar: "Note that the "font" attribute of titleTextAttributes will take precedence over this property"

Comment (jverkoey) [inline components/NavigationBar/src/MDCNavigationBar.h]: There are two things we need to document here:

1. What the default, initial value is.
2. What happens when we set the property to nil.

Comment (mohammadcazig) [reply components/NavigationBar/src/MDCNavigationBar.h]: Done.

Comment (mohammadcazig) [reply components/NavigationBar/src/MDCNavigationBar.h]: Done.

**Review by jverkoey — APPROVED** [review]
